cpp/src/coll/device_kernels.cu

```diff
@@ -0,0 +1,207 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <thrust/functional.h>
+#include <thrust/transform.h>
+
+#include <rmm/exec_policy.hpp>
+
+#include <rapidsmpf/coll/allreduce.hpp>
+#include <rapidsmpf/cuda_stream.hpp>
+#include <rapidsmpf/error.hpp>
+#include <rapidsmpf/memory/buffer.hpp>
+
+namespace rapidsmpf::coll::detail {
+
+using ReduceOperatorFunction = rapidsmpf::coll::ReduceOperatorFunction;
+
+namespace {
+
+template <typename T, typename Op>
+void device_elementwise_reduce(Buffer* acc_buf, Buffer* in_buf, Op op) {
+    RAPIDSMPF_EXPECTS(
+        acc_buf && in_buf, "Device reduction operator requires non-null buffers"
+    );
+    RAPIDSMPF_EXPECTS(
+        acc_buf->mem_type() == MemoryType::DEVICE
+            && in_buf->mem_type() == MemoryType::DEVICE,
+        "Device reduction operator expects device memory"
+    );
+
+    auto const acc_nbytes = acc_buf->size;
+    auto const in_nbytes = in_buf->size;
+    RAPIDSMPF_EXPECTS(
+        acc_nbytes == in_nbytes, "AllReduce device reduction requires equal-sized buffers"
+    );
+    RAPIDSMPF_EXPECTS(
+        acc_nbytes % sizeof(T) == 0,
+        "AllReduce device reduction buffer size must be multiple of sizeof(T)"
+    );
+
+    auto const count = acc_nbytes / sizeof(T);
+
+    // Ensure the accumulator stream waits for the incoming buffer's latest write.
+    cuda_stream_join(acc_buf->stream(), in_buf->stream());
+
+    // Destination buffer: use write_access so the latest-write event is updated.
+    acc_buf->write_access([&](std::byte* acc_bytes, rmm::cuda_stream_view stream) {
+        auto* acc_ptr = reinterpret_cast<T*>(acc_bytes);
+        auto policy = rmm::exec_policy_nosync(stream);
+
+        // Safe to read from incoming buffer after the stream join above.
+        auto const* in_bytes = reinterpret_cast<std::byte const*>(in_buf->data());
+        auto const* in_ptr = reinterpret_cast<T const*>(in_bytes);
+
+        thrust::transform(policy, acc_ptr, acc_ptr + count, in_ptr, acc_ptr, op);
+    });
+}
+
+template <typename T, DeviceReduceOp Op>
+struct ReduceOperatorMaker;
```

Comment:
I prefer operator specialization like this, rather than the `if-else constexpr` in host operators. This is much cleaner.